internal/controller/container_reconciler.go

```diff
@@ -58,6 +63,36 @@ func (r *ContainerImageReconciler) ReconcileContainerImage(ctx context.Context,
 		return result, err
 	}
 
+	if r.Kind == "Notebook" {
+		nb := obj.(*apiv1.Notebook)
+
+		// this is a signed url request for a file with a different md5 checksum
+		if nb.Spec.Image.Upload.Md5Checksum != nb.Status.LastGeneratedMd5Checksum {
+			url, err := r.callSignedUrlGenerator(nb, r.CloudManagerGrpcConn)
+			if err != nil {
+				return result{}, fmt.Errorf("generating upload url: %w", err)
+			}
+
+			nb.Status.UploadURL = url
+			nb.Status.LastGeneratedMd5Checksum = nb.Spec.Image.Upload.Md5Checksum
+			return result{}, nil
+		}
+		if nb.Status.UploadURL != "" {
+			expirationTime, err := r.getExpirationTime(nb.Status.UploadURL)
+			if err != nil {
+				return result{}, fmt.Errorf("getting URL expiration time: %w", err)
+			}
+
+			if time.Now().After(expirationTime) {
+				log.Info("The signed URL has expired. Clearing .Status.UploadURL")
+				nb.Status.UploadURL = ""
+				return result{}, nil
```

Comment:
Hung up here. Unsure how to call Status().Update() properly with these values. It feels like setting `nb.Status.UploadURL` and `nb.Status.LastGeneratedMd5Checksum` in the notebook controller is the right place, but unsure how to bubble that up.

Reply:
Trying to create a `ContainerReconcileResult` type to surface the values to the caller. Standby

Reply:
Tried this direction but it seems like the wrong path. Saved that work on a branch here: https://github.com/substratusai/substratus/commit/04ed77d343fae3e72de0ac4b09f045bd98cdc2b8

Going that route would affect all other controllers that use the ContainerImageReconciler. Doesn't seem correct but maybe.

Still looking for the right way here.

Reply:
I think probably easier to just add a field the container.image.upload dict? So you can stay within the container subspec. That's what Nick does for the container.git as well, he just updated container.name instead of updating a status.

Reply:
Another thing you could do is to change the interface of `ContainerizedObject`. Add the following method to the interface:
```
type ContainerizedObject interface {
	object
	GetImage() *apiv1.Image
        GetStatus() *apiv1.Status # not accurate not sur
}
```

Calling GetStatus would give you a pointer to the Status object that you can then modify inside the ContainerReconciler

Reply:
^ two compelling paths. I'll give these a go after a break. TYSM!